sim/src/main.rs

```diff
@@ -50,40 +50,73 @@ fn main() {
     }
 
     // Create simulator object.
-    let mut simulator: Simulator = Simulator::new(seed);
+    let storage_name = "storage";
+    let mut simulator = sim::simulator::Simulator::new(1); // always run with the seed 0 while we are checking tests
 
-    // node arguments go:  id, capacity, egress_rate, generation_rate, plugin
-    simulator.add_node("traffic-gen", 1, 1, 1, plugin_str);
-    simulator.add_node("node-1", 10, 1, 0, plugin_str);
-    simulator.add_node("node-2", 10, 1, 0, plugin_str);
-    simulator.add_node("node-3", 10, 1, 0, plugin_str);
-    simulator.add_node("node-4", 1, 1, 0, plugin_str); // in setting egress rate to 0, we are making a sink
-    simulator.add_node("node-5", 1, 1, 0, plugin_str); // in setting egress rate to 0, we are making a sink
-    simulator.add_storage("storage");
+    let regular_nodes = [
+        "frontend-v1",
+        "cartservice-v1",
+        "productcatalogservice-v1",
+        "currencyservice-v1",
+        "paymentservice-v1",
+        "shippingservice-v1",
+        "emailservice-v1",
+        "checkoutservice-v1",
+        "recomendationservice-v1",
+        "adservice-v1",
+    ]
+    .to_vec();
+    for node in &regular_nodes {
+        // node: ID, capacity, egress rate, generation rate, plugin
+        simulator.add_node(node, 10, 5, 0, plugin_str);
+    }
+    simulator.add_node("loadgenerator-v1", 10, 1, 1, None);
+    simulator.add_node("sink", 10, 1, 0, None); // rpc sink
+    simulator.add_storage(storage_name);
+
+    // add all connections to storage and to the sink - we want traces to be able to end arbitrarily
+    for node in &regular_nodes {
+        simulator.add_edge(1, node, storage_name, true);
```

Comment:
Does this imply requests can also be sent to storage? This may lead to strange results.

Reply:
What do you mean by requests?  The rpcs have to be able to be sent to storage - that's how storage gets information.

Reply:
The way storage works in the kubernetes env is an off-path request. Storage is not really part of the network (in the sense that requests by clients may travel there). Instead the proxies (and only the proxies) explicitly accesse storage with a custom rpc call.

Reply:
Then I think requests can be sent to storage;  they have to be explicitly addressed to storage though.

Reply:
From the applications point of view storage may as well not exists. But the way this is modelled right now is that every regular node has a link to storage. This seems to apply that RPCs may be sent at random to to storage from these nodes, which is not what we want. This should be controlled by the filter.

Reply:
Would it make more sense to allow edges between plugins and storage?  In that case, the simulator has to be aware of storage as a different entity, because it has to make those edges.  So either the simulator is aware of storage and its special status, or the user has to manually make edges between all the plugins and storage.  I don't like either of those designs but I'm not sure how else one could get a direct link between plugins and storage.

Another option is to have nodes be aware of what storage is, which is similarly untidy.

Reply:
Is it possible to directly access storage from the filter? Logically speaking, this is just an API call after all. Alternatively, is it possible to directly queue requests in storage from a node?

Reply:
```
    auto result = root()->httpCall("storage-upstream",
                                   { {":method", "GET"},
                                    {":path", "/store"},
                                    {":authority", "storage-upstream"},
                                    {"key", key},
                                    {"value", value} },
                                   "", {}, 1000, callback);
```
Looking at the request this is basically a lookup request for the DNS server, which in our case would be a lookup in the map of simulator nodes.

Reply:
Directly from the filter would be a little difficult, I think, because we'd be importing parts of the simulator into the filter.  That seems messy.

What do you mean by directly queue requests in storage from a node?   Right now, we are sending any RPCs from the filter with the destination storage to storage using the node-storage edge.  I don't see how that is any different.  We don't have a lookup in the map of simulator nodes, because we are only allowed to send to nodes that are neighbors.  Although the simulator is slightly more restrictive, I also think it makes sense because we have no application logic, so if we just chose a random node to send everything to, there would be no meaning to the structure of the graph.

Reply:
The graph has meaning to the nodes, but it does not necessarily have to have meaning for things like storage. My initial view was that plugin code could resolve names like "storage" from a central service, get the storage object, and directly query it. But it is probably better to do this over the queue paradigm. Is it possible for the plugin to queue an rpc with destination to storage, which is then picked up by the simulator, disregarding whether there is a connection or not? It should be, right? We just need to remove the check in the plugin wrapper tick. Otherwise I do not see how you can handle this without explicitly attaching storage to every plugin that is created.

Alternatively, you have to implement an explicit check in all nodes to **not** send to storage. Technically, that is also possible since we will implement custom logic for these nodes soon, anyway. Then there will be "random" nodes, which fire RPCs everywhere (including storage), and the bookinfo nodes, which have a specific request/response flow and ignore something like storage.

Reply:
I think it would be possible.  It would mean that we add a destination for every node when they send RPCs off, which isn't a bad idea anyway, if we want to eventually do routing.  

The easiest solution right now is to just add a check saying don't send to storage unless explicitly addressed in all nodes.  But I think the better solution would be to mandate that all RPCs have destinations for the next hop, and sources where they're from.  Nodes would automatically send all RPCs to their plugins upon receiving them, and wait for the next tick to get those RPCs back.  They can tell that an RPC has already been through the filter by the source node.  Then nodes will send to their neighbors.

Edges can pick from their neighbors for a destination, and plugins can choose between the node they're associated with and storage.  Then we can reintroduce plugins as their own elements in the simulator, and instead of making them a field of nodes, we can just make the plugin ID a field of a node, and that's a bit more modular anyway.  In addition, having a plugin will add an extra tick, and simulating some delay time associated with having a filter makes sense anyway.  

I think the only downside of this idea is that we'll have to hard code an ID for storage.  Does that seem like a reasonable solution?

Reply:
I have been thinking about the second option a little, my proposal is the following:
Node is actually a RandomNode, because that is what it does. It sends random requests to anything that is attached. We can keep it as is, and attach storage for now and do not worry about the RPCs it receives. So there are no changes necessary. 

Going forward, when we implement something like bookinfo, we will have to implement custom logic. This includes the destination and request/response patterns. But this logic should only be implemented for the specific node we want to simulate. For example, "productinfo-v1" will send a request to details **and** one of reviews at random. When it gets back a response from either of them, it will send to "loadgen". We can do this in the next pull request.  

Also I assumed that we already have destinations? At least that is what I see from the simulator tick code.

Reply:
That sounds like a good idea.  So when we implement application logic, we just implement another simulator element, and we can have all the nodes, including RandomNode, ProductPage, etc, implement some sort of Node trait.

We have destinations as a "what is the next hop, please go there."  This gets you from a node to an edge, or edge to node.  We do not have routing as in "I am productpage and I want to go to node ratings-v1."  You can't go to a node you do not have a direct connection with, even if it is possible through multiple hops.  However, I think we only need the former, at least for now.